overseer/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+	/// ```
+	/// # use std::time::Duration;
+	/// # use futures::{executor, pin_mut, select, FutureExt};
+	/// # use futures_timer::Delay;
+	/// # use overseer::{
+	/// #     Overseer, Subsystem, SpawnedSubsystem, SubsystemContext,
+	/// #     ValidationSubsystemMessage, CandidateBackingSubsystemMessage,
+	/// # };
+	///
+	/// struct ValidationSubsystem;
+	/// impl Subsystem<ValidationSubsystemMessage> for ValidationSubsystem {
+	///     fn start(
+	///         &mut self,
+	///         mut ctx: SubsystemContext<ValidationSubsystemMessage>,
+	///     ) -> SpawnedSubsystem {
+	///         SpawnedSubsystem(Box::pin(async move {
+	///             loop {
+	///                 Delay::new(Duration::from_secs(1)).await;
+	///             }
+	///         }))
+	///     }
+	/// }
+	///
+	/// struct CandidateBackingSubsystem;
+	/// impl Subsystem<CandidateBackingSubsystemMessage> for CandidateBackingSubsystem {
+	///     fn start(
+	///         &mut self,
+	///         mut ctx: SubsystemContext<CandidateBackingSubsystemMessage>,
+	///     ) -> SpawnedSubsystem {
+	///         SpawnedSubsystem(Box::pin(async move {
+	///             loop {
+	///                 Delay::new(Duration::from_secs(1)).await;
+	///             }
+	///         }))
+	///     }
+	/// }
+	///
+	/// # fn main() { executor::block_on(async move {
+	/// let spawner = executor::ThreadPool::new().unwrap();
+	/// let (overseer, _handler) = Overseer::new(
+	///     Box::new(ValidationSubsystem),
+	///     Box::new(CandidateBackingSubsystem),
+	///     spawner,
+	/// );
+	///
+	/// let timer = Delay::new(Duration::from_secs(1)).fuse();
+	///
+	/// let overseer_fut = overseer.run().fuse();
+	/// pin_mut!(timer);
+	/// pin_mut!(overseer_fut);
+	///
+	/// select! {
+	///     _ = overseer_fut => (),
+	///     _ = timer => (),
+	/// }
+	/// #
+	/// # }); }
+	/// ```
+	pub fn new(
+		validation: Box<dyn Subsystem<ValidationSubsystemMessage> + Send>,
+		candidate_backing: Box<dyn Subsystem<CandidateBackingSubsystemMessage> + Send>,
+		mut s: S,
+	) -> (Self, OverseerHandler) {
+		let (events_tx, events_rx) = mpsc::channel(CHANNEL_CAPACITY);
+
+		let handler = OverseerHandler {
+			events_tx: events_tx.clone(),
+		};
+
+		let mut running_subsystems_rx = StreamUnordered::new();
+		let mut running_subsystems = FuturesUnordered::new();
+
+		let validation_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			validation,
+		);
+
+		let candidate_backing_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			candidate_backing,
+		);
+
+		let this = Self {
+			validation_subsystem,
+			candidate_backing_subsystem,
+			s,
+			running_subsystems,
+			running_subsystems_rx,
+			events_rx,
+		};
+
+		(this, handler)
+	}
+
+	// Stop the overseer.
+	async fn stop(mut self) {
+		if let Some(ref mut s) = self.validation_subsystem.instance {
+			let _ = s.tx.send(FromOverseer::Signal(OverseerSignal::StopWork)).await;
```

Comment:
`StopWork` is supposed to be about concluding work on a particular relay-parent, so we should introduce a new `Conclude` variant of `OverseerSignal` for telling subsystems to fully conclude.

Reply:
added